src/main/java/org/pesmypetcare/webservice/dao/usermanager/UserDaoImpl.java

```diff
@@ -0,0 +1,499 @@
+
+    @Override
+    public UserEntity getUserData(String uid) throws DatabaseAccessException {
+        DocumentSnapshot userDoc = getDocumentSnapshot(users, uid);
+        throwExceptionIfUserDoesNotExist(userDoc);
+        return userDoc.toObject(UserEntity.class);
+    }
+
+    @Override
+    public void updateField(String username, String field, String newValue)
+        throws FirebaseAuthException, DatabaseAccessException {
+        String uid = getUid(username);
+        switch (field) {
+            case USERNAME_FIELD:
+                updateUsername(uid, newValue);
+                break;
+            case EMAIL_FIELD:
+                updateEmail(uid, newValue);
+                break;
+            case PASSWORD_FIELD:
+                updatePassword(uid, newValue);
+                break;
+            default:
+                throw new DatabaseAccessException("invalid-field", "The field does not exist");
+        }
+    }
+
+    @Override
+    public boolean existsUsername(String username) throws DatabaseAccessException {
+        DocumentSnapshot usernameDoc = getDocumentSnapshot(usedUsernames, username);
+        return usernameDoc.exists();
+    }
+
+    @Override
+    public String getField(String uid, String field) throws DatabaseAccessException {
+        DocumentSnapshot userDoc = getDocumentSnapshot(users, uid);
+        throwExceptionIfUserDoesNotExist(userDoc);
+        return (String) userDoc.get(field);
+    }
+
+    @Override
+    public String getUid(String username) throws DatabaseAccessException {
+        DocumentSnapshot usernameDoc = getDocumentSnapshot(usedUsernames, username);
+        throwExceptionIfUserDoesNotExist(usernameDoc);
+        return (String) usernameDoc.get(USER_KEY);
+    }
+
+    @Override
+    public void addGroupSubscription(String username, String groupName, WriteBatch batch)
+        throws DatabaseAccessException {
+        DocumentReference user = users.document(getUid(username));
+        Map<String, Object> data = new HashMap<>();
+        data.put("groupSubscriptions", FieldValue.arrayUnion(groupName));
+        batch.update(user, data);
+    }
+
+    @Override
+    public void deleteGroupSubscription(String userUid, String groupName, WriteBatch batch) {
+        DocumentReference user = users.document(userUid);
+        Map<String, Object> data = new HashMap<>();
+        data.put("groupSubscriptions", FieldValue.arrayRemove(groupName));
+        batch.update(user, data);
+    }
+
+    @Override
+    public List<String> getUserSubscriptions(String username) throws DatabaseAccessException {
+        DocumentSnapshot user = getDocumentSnapshot(users, getUid(username));
+        return (List<String>) user.get("groupSubscriptions");
+    }
+
+    @Override
+    public void saveMessagingToken(String uid, String token) throws DatabaseAccessException, DocumentException {
+        String currentToken = documentAdapter.getStringFromDocument(Path.ofDocument(Collections.users, uid), "FCM");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The String "FCM" appears 4 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5561439)